In [33]:
!pip install imbalanced-learn

In [39]:
from imblearn.under_sampling import RandomUnderSampler, ClusterCentroids

In [35]:
import imblearn

In [36]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from imblearn.under_sampling import RandomUnderSampler, TomekLinks
from imblearn.over_sampling import RandomOverSampler, SMOTE

Step 2:Reading the dataset

In [37]:
dataset = pd.read_csv('/content/creditcard.csv')
print("The Number of Samples in the dataset: ", len(dataset))
print('Class 0         :',round(dataset['Class'].value_counts()[0]
                        /len(dataset)   * 100,2), '% of the dataset')
print('Class 1(Fraud) :', round(dataset['Class'].value_counts()[1]
                                / len(dataset)   * 100, 2), '% of the dataset')

The Number of Samples in the dataset:  284807
Class 0         : 99.83 % of the dataset
Class 1(Fraud) : 0.17 % of the dataset


Step 3: understanding of major class

In [38]:
#we will do an undersampling of the major class where the cardholder is not fraud through this technique we will reduce the rows which come under the major class
X_data = dataset.iloc[:, :-1]
Y_data = dataset.iloc[:, -1:]

rus = RandomUnderSampler(random_state=42)

X_res, y_res = rus.fit_resample(X_data, Y_data)

X_res = pd.DataFrame(X_res)
Y_res = pd.DataFrame(y_res)

print("After Under Sampling Of Major Class Total Samples are :", len(Y_res))
print('Class 0          :', round(Y_res.value_counts()/
                                  [0] / len(Y_res) * 100, 2), '% of the dataset')
print('Class 1 (Fraud)    :', round(Y_res.value_counts()\
                                    [1] / len(Y_res) * 100, 2), '% of the dataset')


After Under Sampling Of Major Class Total Samples are : 984
Class 0          : Class
0        inf
1        inf
dtype: float64 % of the dataset
Class 1 (Fraud)    : 50.0 % of the dataset


We can see after doing undersampling the total data in the major class has reduced to 984

Step4:Undersampling Using Tomelinks

In [32]:
tl = TomekLinks()

X_res, y_res = tl.fit_resample(X_data, Y_data)
X_res = pd.DataFrame(X_res)
Y_res = pd.DataFrame(y_res)


print("After TomiLinks Under Sampling Of Major\
          Class Total Samples are:", len(Y_res))
print('Class 0   :', round(Y_res.value_counts()[0] / len(Y_res) * 100,2), '% of the dataset')
print('Class 1(Fraud)  :', round(Y_res.value_counts()[1] / len(Y_res) * 100,2), '% of the dataset')

After TomiLinks Under Sampling Of Major          Class Total Samples are: 284736
Class 0   : 99.83 % of the dataset
Class 1(Fraud)  : 0.17 % of the dataset


Step 5:Oversampling Using RandomOversampler
We can use RandomOversampler to oversample the minority class data.Using Random Oversample the model picks randomly data points from the existing datasets

In [43]:
ros = RandomOverSampler(random_state=42)

X_res, y_res = ros.fit_resample(X_data, Y_data)
X_res = pd.DataFrame(X_res)
Y_res = pd.DataFrame(y_res)

print("After Over Sampling Of Minor Class\
              Total Samples are :", len(Y_res))
print('Class 0      :',round(Y_res.value_counts()[0]   / len(Y_res)  * 100,2),  '% of the dataset')
print('Class 1(Fraud)  :', round(Y_res.value_counts()[1]   / len(Y_res)  * 100,2),  '% of the dataset')


After Over Sampling Of Minor Class              Total Samples are : 568630
Class 0      : 50.0 % of the dataset
Class 1(Fraud)  : 50.0 % of the dataset


Step6:Oversampling UsingSMOTE
We can use Smote to generate random sample data for the minority class.One useful thing about using SMOTE is that it creates new synthic data points for the minority class

In [45]:
sm = SMOTE(random_state=42)

X_res, y_res = sm.fit_resample(X_data, Y_data)

X_res = pd.DataFrame(X_res)
Y_res = pd.DataFrame(y_res)

print("After SMOTE Over Sampling Of Minor\
               Class Total Samples are :", len(Y_res))
print('Class 0       :', round(Y_res.value_counts()[0]  / len(Y_res)   * 100, 2), '% of the dataset')
print('Class 1(Fraud)    :', round(Y_res.value_counts()[1]   / len(Y_res)  * 100, 2), '% of the dataset')

After SMOTE Over Sampling Of Minor               Class Total Samples are : 568630
Class 0       : 50.0 % of the dataset
Class 1(Fraud)    : 50.0 % of the dataset
